crates/miden-tx/src/executor/notes_checker.rs

```diff
@@ -62,30 +33,56 @@ impl NotesChecker {
     #[maybe_async]
     pub fn check_notes_consumability(
         &self,
-        tx_executor: &TransactionExecutor,
+        target_account_id: AccountId,
         block_ref: BlockNumber,
+        input_notes: InputNotes<InputNote>,
         tx_args: TransactionArgs,
     ) -> Result<NoteAccountExecution, TransactionExecutorError> {
         // Check input notes
         // ----------------------------------------------------------------------------------------
-        let inputs_check_result = self.check_note_inputs();
-        if let NoteInputsCheck::No { failed_note_id } = inputs_check_result {
-            return Ok(NoteAccountExecution::Failure {
-                failed_note_id,
-                successful_notes: vec![],
-                error: None,
-            });
+
+        let mut successful_notes = vec![];
+        for note in input_notes.iter() {
```

Comment:
It seems like the general approach of the code is now to do as many checks in Rust as possible. So I wonder if we should loop over all notes here and check all the P2ID(R) notes, regardless of whether we have encountered another note yet. Basically, the idea being that if there is a transaction with a SWAP and P2ID note (in that order) but the note inputs of the P2ID note are incorrect, then we can already abort at this stage without ever having to execute the transaction in the VM.

So maybe we can just introduce a flag `run_check_in_vm = false` which is set to `true` if a non-P2ID(R) note is encountered in the loop, but we finish the loop anyway? And then we only run `try_execute_notes` if that flag is `true` after the loop.

Reply:
I agree, that will be much better!

Reply:
This comment hasn't been addressed yet, right?

Reply:
I addressed it, but in a slightly different way. I didn't introduce the flag, but used the `successful_notes` vector to check whether we need to run the transaction (and I changed `break`s to the `continue`s to check all the notes in any case). Potentially we could get rid of this vector, but in that case we won't be able to return a list of successful notes if the check returns `No`.